pool/service.go

```diff
@@ -243,19 +287,51 @@ func (p *VipnodePool) Client(ctx context.Context, sig string, nodeID string, non
 		return nil, err
 	}
 
-	r, err := p.Store.ActiveHosts(kind, numRequestHosts)
+	reqKind := kind
+	if isHost {
+		// Any kind of host peer will do.
+		reqKind = ""
+	}
+
+	hosts, err := p.requestHosts(ctx, nodeID, req.NumHosts, reqKind)
 	if err != nil {
 		return nil, err
 	}
-	if len(r) == 0 {
-		logger.Printf("New %q client: %q (no active hosts found)", kind, pretty.Abbrev(nodeID))
+
+	if !isHost && len(hosts) == 0 {
+		logger.Printf("New %q peer: %q (no active hosts found)", kind, pretty.Abbrev(nodeID))
 		return nil, NoHostNodesError{}
 	}
 
+	response.Hosts = hosts
 	if p.skipWhitelist {
-		logger.Printf("New %q client: %q (%d hosts found, skipping whitelist)", kind, pretty.Abbrev(nodeID), len(r))
-		response.Hosts = r
-		return response, nil
+		logger.Printf("New %q peer: %q (%d hosts found, skipping whitelist)", kind, pretty.Abbrev(nodeID), len(hosts))
+	} else {
+		logger.Printf("New %q peer: %q (%d hosts found)", kind, pretty.Abbrev(nodeID), len(hosts))
+	}
+
+	return response, nil
+}
+
+func (p *VipnodePool) requestHosts(ctx context.Context, nodeID string, numRequestHosts int, kind string) ([]store.Node, error) {
```

Comment:
RFC: Is it worth adding a RequestHosts pool RPC? Right now it's an implicit part of Connect, but it could make sense to just call RequestHosts periodically out of sync.

(Is there a better name for this? `vipnode_requestHosts`?)

Reply:
Are vipnode members currently only told about hosts on connect?  Is there a way to push/poll for new hosts?  If not then ya either the vipnode client could poll for hosts via a `vipnode_requestHosts` RPC, or new hosts could be pushed down via JSONRPC notifications.

I think the important of this depends on whether or not both hosts in a pool need to mutually AddTrustedPeer each other before they have a decent chance of connecting to one another, or if just one side AddTrustedPeer'ing the other is typically enough.

If both sides need to trust the other, then when a new host joins the pool unless the other hosts are told about it by some mechanism it seems like that new host won't be as well peered as the existing ones.

Reply:
Alternatively, what if `vipnode_update` responses included a recent list of hosts?

Reply:
Keep in mind that requestHosts doesn't just return random hosts willy-nilly, but it goes through the trouble of asking the hosts to pre-emptively whitelist the requester's nodeID, and waits for them to return successfully.

The number of pre-whitelisted hosts you connect to will effectively increase your cost burden as a light client, since you're metered per connection pair. If you're connected to 1 premium host, you're paying $1*X/minute. If you're connected to 5 premium hosts, you're paying $5*X/minute.

While it's possible for the pool to provision more host peers, I think it makes sense for the client agent to semi-explicitly request more pre-whitelisted hosts, rather than the pool always just feeding in hosts as it feels like.

From the agent's perspective, I already added a flag to request a specific number of hosts, so we can setup the agent to request more if the number drops below that.

I think the options for now are:
1. Keep it as-is and only request hosts on-connect. If you need more hosts, reconnect.
2. Request hosts on-connect, _and_ request more later through a new RPC.
3. Don't even requests hosts on-connect, but automatically request hosts through the new RPC immediately after connecting.

I'm leaning towards 3.

Also is there a better name than `vipnode_requestHosts`?

Reply:
```go
type Pool interface {
    ...

	RequestHosts(ctx context.Context, req RequestHostsRequest) (*RequestHostsRequest, error)
}
```

lol `RequestHostsRequest`. Surely there must be a better name... :thinking:

Reply:
> Keep in mind that requestHosts doesn't just return random hosts willy-nilly, but it goes through the trouble of asking the hosts to pre-emptively whitelist the requester's nodeID, and waits for them to return successfully.

I think the key to a better name lies in this statement.  Some ideas:

- `vipnode_requestHostPeers`
- `vipnode_getHostsToPeerWith`
- `vipnode_initiateHostPeering`
- `vipnode_startHostPeerHandshake`

The point being to focus on _why_ the hosts are being requested.

Reply:
`vipnode_gimmeHostsWhoAlreadyWhitelistedMe` :P 

But seriously, leaning on `vipnode_getHosts`. I can imagine a scenario in the future where the pre-whitelisted or not is a property of the request payload.